assembl/models/idea.py

```diff
@@ -999,11 +999,30 @@ class RootIdea(Idea):
 
     @property
     def num_posts(self):
-        """ In the root idea, this is the count of all mesages in the system """
+        """ In the root idea, num_posts is the count of all non-deleted mesages in the discussion """
+        from .post import Post
+        result = self.db.query(Post).filter(
+            Post.discussion_id == self.discussion_id,
+            Post.hidden==False,
+            Post.tombstone_condition()
+        ).count()
+        return int(result)
+
+    @property
+    def num_read_posts(self):
+        """ In the root idea, num_posts is the count of all non-deleted read mesages in the discussion """
         from .post import Post
```

Comment:
It not always faster or slower.  Like many things in life, it depends ;)

Import in the function body is going to be faster (and possibly consume a lot les memory) if it avoids some modules being loaded (lazy loading modules) but significantly slower if the function is small and executed in a tight loop (there is function call overhead from the import statement even if the module is loaded only once).

In any event, we did it to avoid circular loading.  Sqlalchemy declarative models requires some hard to predict module loading orders.

Should we decide to make it a standard to not load module in function calls, we must move ALL imports to the end, otherwise we will have NEITHER the readability/documentation advantages of grouped imports, OR the lazy loading/documentation locality advantages of in function imports.